monai/handlers/checkpoint_loader.py

```diff
@@ -53,6 +54,11 @@ class CheckpointLoader:
             This can be useful advanced feature for transfer learning. users should totally
             understand which layers will have different shape. default to `True`.
 
+    Note: if `strict_shape=False`, will only load checkpoint for `torch.nn.Module` and skip other
```

Comment:
This flag is a bit confusing, I think the name could be "nn_modules_only" or "nn_mod_only"? But if this is the case, do we support nn.module in a nested dict, such as {"models": {"test": {"model": nn_parameters}}}

Reply:
Hi @wyli , the main point here is that: `strict_shape=False` can only work with model, if model layers have different shape, we should not load `state_dict` for the related optimizer and lr_sheduler, etc. anymore.
What do you think a better name for this arg?

Thanks.

Reply:
I'm not sure if I understand the current `strict` and `strict_shape` correctly, could you document more in the unit test for the test cases, or @sandylaker could you please help review this pull request?

Reply:
Hi @wyli ,

Sure, I enhanced the unit tests with more check and doc-strings.
Actually, I think if loading checkpoint for a model with different shape layers, usually it's not a continuous training, so should not load the checkpoint of optimizer, lr_scheduler, etc.
And if model **has different shape layers**, we should not load other related things, which may cause an unexpected error as the `state_dict` of these components are not pure `Tensor` like nn.Module.
What do you think?

Thanks.

Reply:
@wyli 

1. The `strict` flag checks if the model's keys match the state dictionary's keys. Here the state dictionary is loaded from the check point. For example, If `model = nn.Linear(2, 1, bias=False)` and the `state_dict` to be loaded is from `nn.Linear(2, 1, bias=True)`, then setting `strict = False` allows for loading such `state_dict`, otherwise `torch` will complain about unmatched keys, see [here](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.load_state_dict). It is also to notice that this flag is actually passed to `ignite.handlers.Checkpoint.load_state_dict` (see [here](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.load_state_dict)) at the first place. In this function, `strict` has only effect to `nn.Module`, `DP` and `DDP` wrappers. For other classes like `Optimizer`, `_LRscheduler`, their `load_state_dict` does not accept extra flags. In this case, the `strict` flag is simply ignored.
2. The `strict_shape` forces the tensors of matched keys to have the same shape. Thus, when doing transfer learning, it is better to set `strict_shape=False`. E.g., one might hope to fine-tune a model on CIFAR-10 with pre-trained weights on ImageNet. In this case, the shape of last linear layer is different.

To make the docs more comprehensive, it would be great if a link to `ignite.Checkpoint.load_state_dict` or `nn.Module.load_state_dict` is added. Thus, the user can refer to these links when confused.

Reply:
Hi @sandylaker ,

Thanks for your detailed explanation!
Sure, I added the 2 links in the latest commit.

Thanks.

Reply:
thanks @sandylaker, @Nic-Ma, I was confused by two things, 
1. "model keys" are the keys in `dict(nn.modules.named_parameters())`, not the keys of `load_dict` such as `load_dict={"net": net2, "opt": opt2}`, 

2. in the current implementation,  `strict_shape=False` is actually implemented as "don't check anything", the `strict` flag is ignored... I thought `strict_shape=False` means the keys should match, but allow for shape mismatch.

Reply:
Hi @wyli ,

1. Sure, let me try to enhance the doc-string to make the `model keys` more clear in `strict` arg.
2. Actually, we can't allow "keys must match but shape mismatch", because the `strict` check for keys is in PyTorch itself when **applying** checkpoint, if the shape mismatch, it will throw an exception. So can we enhance the doc-string of `strict_shape` to make it clear to users that it will ignore `strict=True`?

Thanks.

Reply:
I think you put a warning there already. Maybe it's fine for now

Reply:
Enhanced doc-strings in the latest commit.
Thanks.